Enable Fully Concurrent Snapshot Operations

original-brownbear opened the pull request:
Enables fully concurrent snapshot operations:
* Snapshot create- and delete operations can be started in any order
* Delete operations wait for snapshot finalization to finish, are batched as much as possible to improve efficiency and once enqueued in the cluster state prevent new snapshots from starting on data nodes until executed
   * We could be even more concurrent here in a follow-up by interleaving deletes and snapshots on a per-shard level. I decided not to do this for now since it seemed not worth the added complexity yet. Due to batching+deduplicating of deletes the pain of having a delete stuck behind a long -running snapshot seemed manageable (dropped client connections + resulting retries don't cause issues due to deduplication of delete jobs, batching of deletes allows enqueuing more and more deletes even if a snapshot blocks for a long time that will all be executed in essentially constant time (due to bulk snapshot deletion, deleting multiple snapshots is mostly about as fast as deleting a single one))
* Snapshot creation is completely concurrent across shards, but per shard snapshots are linearized for each repository as are snapshot finalizations

See updated JavaDoc and added test cases for more details and illustration on the functionality.

Some notes:

The queuing of snapshot finalizations and deletes and the related locking/synchronization is a little awkward in this version but can be much simplified with some refactoring.  The problem is that snapshot finalizations resolve their listeners on the `SNAPSHOT` pool while deletes resolve the listener on the master update thread. With some refactoring both of these could be moved to the master update thread, effectively removing the need for any synchronization around the `SnapshotService` state. I didn't do this refactoring here because it's a fairly large change and not necessary for the functionality but plan to do so in a follow-up.

This change allows for completely removing any trickery around synchronizing deletes and snapshots from SLM and 100% does away with SLM errors from collisions between deletes and snapshots.

Snapshotting a single index in parallel to a long running full backup will execute without having to wait for the long running backup as required by the ILM/SLM use case of moving indices to "snapshot tier". Finalizations are linearized but ordered according to which snapshot saw all of its shards complete first

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

original-brownbear committed:
Fully Concurrent Snapshots

Implements fully concurrent snapshot operations. See documentation changes
to snapshot package level JavaDoc for details.

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/MinThreadsSnapshotRestoreIT.java:
```diff
@@ -1,155 +0,0 @@
- * in the thread pool (for example, tests that use the mock repository that
- * block on master).
- */
-public class MinThreadsSnapshotRestoreIT extends AbstractSnapshotIntegTestCase {
```
All scenarios covered by this test become obsolete. The actual premise of this test (checking that we don't dead-lock from blocked threads) is covered by the fact that `SnapshotResiliencyTests` work for the most part anyway.

original-brownbear committed:
cleanup todo

original-brownbear committed:
docs

original-brownbear committed:
cleanup

original-brownbear committed:
doc

original-brownbear committed:
reduce noise

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -232,7 +232,7 @@ private void startNewSnapshots(SnapshotsInProgress snapshotsInProgress) {
                         // due to CS batching we might have missed the INIT state and straight went into ABORTED
                         // notify master that abort has completed by moving to FAILED
-                        if (shard.value.state() == ShardState.ABORTED) {
+                        if (shard.value.state() == ShardState.ABORTED && localNodeId.equals(shard.value.nodeId())) {
```
This fixes a tricky bug that might have been trouble for aborts with pre-7.6 repository metadata on S3 and generally wastes a lot of shard status updates. I'll fix this in a separate PR right away, just adding a node here to not forget to handle that first.

original-brownbear replied:
https://github.com/elastic/elasticsearch/pull/58214

original-brownbear committed:
less noise

original-brownbear committed:
fix test

original-brownbear committed:
cover primary fail-over

original-brownbear committed:
drier tests

original-brownbear committed:
fix test

original-brownbear committed:
update comment

original-brownbear committed:
drier

original-brownbear committed:
make things nicer looking

original-brownbear committed:
drop pointless short-circuit

original-brownbear committed:
doc

original-brownbear committed:
moar test

original-brownbear committed:
another corner case fixed

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
more docs

original-brownbear committed:
bck

original-brownbear committed:
save some more repo data loading

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
moar corner cases

original-brownbear committed:
cleanup tests more

original-brownbear committed:
cleanups

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -725,10 +857,95 @@ private void handleFinalizationFailure(Exception e, SnapshotsInProgress.Entry en
+     * Runs a cluster state update that checks whether we have any waiting snapshot deletions that can be executed and if so sets them
+     * to state {@link SnapshotDeletionsInProgress.State#META_DATA} and executes them.
+     *
+     * TODO: optimize this to execute in a single CS update together with finalizing the latest snapshot
```
This is something to maybe handle before releasing (so that we never get a waiting delete that could be executed on master fail-over and can simplify the logic htere) but it requires non-trivial refactoring to make this happen and this PR is already huge and I figured it's better to get alignment on the strategy of how to merge this before doing that kind of refactoring.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1441,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
-                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", snapshotIds), e);
-                if (listener != null) {
-                    listener.onFailure(e);
+                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", deleteEntry), e);
```
This is a scenario that was complicated before but became very complicated now, both fail-over on delete and create. We could put more effort into this by doing things like https://github.com/elastic/elasticsearch/pull/54350 but I think it's not a priority relative to getting concurrent operations work functionally so I just went with the brute-force fail everything on master fail-over approach here as that's the current behavior for single snapshots.

original-brownbear commented on server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java:
```diff
@@ -594,11 +582,16 @@ public void testBulkSnapshotDeleteWithAbort() {
                         .setWaitForCompletion(true).execute(createSnapshotResponseStepListener));
 
-        final StepListener<CreateSnapshotResponse> createOtherSnapshotResponseStepListener = new StepListener<>();
+        final int inProgressSnapshots = randomIntBetween(1, 5);
```
Here it was easy to expand coverage but unfortunately `SnapshotResiliencyTests` don't really cover repository exception handling so that's where `ConcurrentSnapshotITs` come from.

original-brownbear replied:
`IOExceptions` thrown by the repository. We don't have any simulation of that in `SnapshotResiliencyTests`

original-brownbear committed:
be smarter about failover handling

original-brownbear committed:
fewer CS updates

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
nicer

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/ConcurrentSnapshotsIT.java:
```diff
@@ -0,0 +1,889 @@
+import static org.hamcrest.Matchers.not;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class ConcurrentSnapshotsIT extends AbstractSnapshotIntegTestCase {
```
I know this class is massive, but I did my best to get full coverage of all possible permutations of master fail-over and repository IO issues here (we definitely had some gaps in our testing without it even for single snapshots as this one caught #58214).

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -176,17 +231,23 @@ public String toString() {
+        private final State state;
         private final long startTime;
         private final long repositoryStateId;
+        private final String uuid;
```
Needed the new id here to get some fixed coordinate/key for a delete operation to connect completion listeners to when batching deletes up.

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -176,17 +231,23 @@ public String toString() {
     public static final class Entry implements Writeable, RepositoryOperation {
         private final List<SnapshotId> snapshots;
         private final String repoName;
+        private final State state;
```
Needed to know whether or not a delete can be executed yet or is still open to have snapshot ids added to it. Somewhat analogous to snapshot in progress state.
Technically, a boolean would be enough here. But I'd like to keep that state enum here to maybe further enhance the state machine with a cleanup step. As an improvement for concurrent operations, we could already move on to the next snapshot operation after master is done writing the new metadata during a delete but before it deleted all the unreferrenced blobs. It would be nice for the cleanup to still be recorded in the cluster state though as otherwise back-to-back deletes might list out and delete the same blobs concurrently+redundantly. It would just be easier to create that follow-up if we didn't go with a flag here now.

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -76,13 +92,52 @@ public SnapshotDeletionsInProgress withAddedEntry(Entry entry) {
+        return removed ? new SnapshotDeletionsInProgress(updatedEntries) : this;
+    }
+
+    public SnapshotDeletionsInProgress withRemovedSnapshotIds(String repository, Collection<SnapshotId> snapshotIds) {
```
If two or more deletes share overlapping snapshots but couldn't be batched into one we need to "clean up" deleted snapshots from existing deletes so that the remaining snapshots don't fail trying to delete already deleted snapshots.

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -76,13 +92,52 @@ public SnapshotDeletionsInProgress withAddedEntry(Entry entry) {
+        return changed ? new SnapshotDeletionsInProgress(updatedEntries) : this;
+    }
+
+    public SnapshotDeletionsInProgress withRemovedRepository(String repository) {
```
If we fail loading repository data for a repository but have queued up deletes we just drop all the deletes for the repo on failure.

original-brownbear replied:
No if we do this in a CS update we collect the relevant entries in the CS update and fail all the listeners for which we removed the job from the state.
See here https://github.com/elastic/elasticsearch/pull/56911/files#diff-a0853be4492c052f24917b5c1464003dR1512

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1616,6 +1615,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     * @param newGen new safe repository generation
+     * @return updated cluster state
+     */
+    private ClusterState updateRepositoryGenerations(ClusterState state, long oldGen, long newGen) {
```
It's somewhat questionable to even continue tracking the generation of an operation at this point since it's always equal to the safe repository generation (except for the corner case of having just mounted the repository maybe ...). I added the generation updating here for now to keep the change-set small but I'd look into removing the generation from deletes and snapshots in a follow-up (technically easy ... but BwC of that change is tricky).

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1046,25 +1247,76 @@ public ClusterState execute(ClusterState currentState) {
+                                    final ImmutableOpenMap<ShardId, ShardSnapshotStatus> abortedShards = abortEntry(existing);
+                                    final boolean isCompleted = completed(abortedShards.values());
+                                    final SnapshotsInProgress.Entry abortedEntry = new SnapshotsInProgress.Entry(
+                                            existing, isCompleted ? State.SUCCESS : State.ABORTED, abortedShards,
```
This is somewhat stupid, but I made this work the same way shard state updates are processed right now (setting success if all the shards completed). A follow-up to simplify and speed things up would be to simply remove snapshots that haven't done any work yet here right away instead of going through the redundant cycle of finalizing them and then deleting them again right away.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
                     }
                     if (changed) {
-                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions).build();
+                        // We removed a delete from the cluster state. If that delete caused any snapshots to not have their shard
```
This is fairly complex code now unfortunately since we have to deal with assigning new work where possible across queued up snapshots, truncate queued up deletes and handle the complex failure situation of having failed to read repository data before executing the delete.
I did my best to comment reasonably and tried to reuse the shard assignment logic used during snapshot create to keep it simple instead of optimal (see added TODO about optimizing this logic).

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/Repository.java:
```diff
@@ -152,7 +152,7 @@ void finalizeSnapshot(SnapshotId snapshotId, ShardGenerations shardGenerations,
      */
     void deleteSnapshots(Collection<SnapshotId> snapshotIds, long repositoryStateId, Version repositoryMetaVersion,
-                         ActionListener<Void> listener);
+                         ActionListener<RepositoryData> listener);
```
Same as with snapshot finalizations, much easier to keep things consistent (and faster) by passing the repository data through the listeners instead of having to reload it over and over.

original-brownbear commented:
@ywelsch @tlrx I think this one is good to go at last now. Unfortunately with the introduction of concurrent delete operations this went from 500 lines to 2k lines (a large chunk of it in new tests) and added a lot of complex logic for handling all the various corner cases around aborting snapshots etc. 

I could split this up in two pieces by introducing concurrent deletes first (that would separate out all the changes to the delete state, batching up of delete operations and so on) if that helps the review. I just figured the delete changes in isolation are not easy to judge without having the full picture on why I made the choices I made in this PR so I went with the full end-to-end solution for starters.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
reproducer

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -725,12 +836,107 @@ private void handleFinalizationFailure(Exception e, SnapshotsInProgress.Entry en
+        });
+    }
+
+    private static Tuple<ClusterState, List<SnapshotDeletionsInProgress.Entry>> readyDeletions(ClusterState currentState) {
```
add javadocs here to explain what this method does?

original-brownbear replied:
Done :)

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -821,22 +1023,23 @@ public void deleteSnapshots(final DeleteSnapshotRequest request, final ActionLis
                     return deleteFromRepoTask.execute(currentState);
                 }
-
+                assert snapshotEntries.size() == 1 : "Expected just a single running snapshot but saw " + snapshotEntries;
```
why is there only at most one?

original-brownbear replied:
Because we break out above to the new logic if we have a cluster on or after `FULL_CONCURRENCY_VERSION` and hence never get here unless there  is precisely one running snapshot.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        if (failure == null) {
+                            // The delete worked out so we remove the snapshot ids that it removed from the repository from queued up
+                            // delete jobs
+                            deletions = deletions.withRemovedSnapshotIds(deleteEntry.repository(), deleteEntry.getSnapshots());
```
do we need to set changed to true here?

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        if (failAllQueuedOperations) {
+                            deletions.getEntries().stream().filter(entry -> entry.repository().equals(deleteEntry.repository()))
+                                    .map(SnapshotDeletionsInProgress.Entry::uuid).forEach(deletionsToFail::add);
+                            // We failed to read repository data for this delete, it is not the job of SnapshotsService to
```
this should only be a problem if the repository data isn't cached, right?

original-brownbear replied:
Yea it's a highly unlikely scenario, that's why I figured brute force failing (just removing from CS) all things queued up for a given snapshot should be ok here.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                                    .map(SnapshotDeletionsInProgress.Entry::uuid).forEach(deletionsToFail::add);
+                            // We failed to read repository data for this delete, it is not the job of SnapshotsService to
+                            // retry this kind of issue so we fail all the pending deletes
+                            deletions = deletions.withRemovedRepository(deleteEntry.repository());
```
Should there be a return after this line?

I see usage later on of repositoryData where it's possibly null

ywelsch replied:
Also sounds like we have a testing gap

original-brownbear replied:
I think we're safe here because we don't do any actual shard assignments that would use `RepositoryData` if it's `null`, we just break out here: https://github.com/elastic/elasticsearch/pull/56911/files#diff-a0853be4492c052f24917b5c1464003dR1529 and fail all the queued up snapshots and deletes. This is actually covered by e.g. test `org.elasticsearch.snapshots.ConcurrentSnapshotsIT#testQueuedDeletesWithFailures` that has both a queued up snapshot and delete fail as a result of not being able to read `RepositoryData`

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                                    } else if (entry.state().completed() == false) {
+                                        boolean updatedQueuedSnapshot = false;
+                                        for (ObjectCursor<ShardSnapshotStatus> value : entry.shards().values()) {
+                                            if (value.value.equals(ShardSnapshotStatus.UNASSIGNED_WAITING)) {
```
As mentioned in Slack, this could possibly be result in shards to be done out-of-snapshot-order in case where there is a relocating primary.

original-brownbear replied:
Fixed this case now in https://github.com/elastic/elasticsearch/pull/56911/commits/ee8827588c76bba0ab2181691a99027aa6df90be and added test+assertion coverage for it.

:100: thanks for spotting this!

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        assert snapshotDeletionListeners.isEmpty() :
+                                "No new listeners should have been added but saw " + snapshotDeletionListeners;
+                    } else {
+                        assert false : "Removing snapshot entry should only ever fail because we failed to publish new state";
```
should we still call the listeners here?

original-brownbear replied:
I don't know. It's a broken situation either way if we get here and come to think of it we have the same issue in the snapshot path. In the end, only a master fail-over can fix things here and it should be impossible to arrive here unless there is a bug.
If we call the listeners we'll return from the request but the cluster state is still stuck. If we don't call the listeners we won't return from the request so at least the issue is a little more visible? Seems to me it makes no difference what we do here practically.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                    failListenersIgnoringException(deleteListeners, failure);
+                }
+                runningDeletions.remove(deleteEntry.uuid());
+                if (repositoryData == null) {
```
how is this tested?

original-brownbear replied:
`org.elasticsearch.snapshots.ConcurrentSnapshotsIT#testQueuedOperationsAndBrokenRepoOnMasterFailOver` reproduces this scenario by corrupting the index-N blob before master fail-over.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1322,20 +1811,41 @@ protected void doClose() {
+                        } else {
+                            final Set<ShardId> reusedShardIds =
+                                    reusedShardIdsByRepo.computeIfAbsent(entry.repository(), k -> new HashSet<>());
+                            if (entry.state().completed() == false && reusedShardIds.contains(finishedShardId) == false
```
Does this logic still make sure that a snapshot that's taken contains a set of docs that existed sometime between start and end of snapshot command? Or will this batch up snapshots that had been running prior to the command?

original-brownbear replied:
> Does this logic still make sure that a snapshot that's taken contains a set of docs that existed sometime between start and end of snapshot command? 

Yes. There is no smart logic here. We simply reassign the finished shards as (new `INIT` state shard snapshots) if there's other outstanding snapshots for them and put them into `reusedShardIds` to ensure we only reassign once per shard+repo => a new snapshot will be taken from a fresh index commit so the guarantees about the timing are unchanged relative to what we have right now.

ywelsch reviewed (changes requested):
I've done an initial pass, but can't claim to fully understand everything just yet.

original-brownbear committed:
fix waiting+relocation failure propagation

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
docs fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -666,8 +753,36 @@ private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata) {
             return;
         }
         final Snapshot snapshot = entry.snapshot();
+        if (entry.repositoryStateId() == RepositoryData.UNKNOWN_REPO_GEN) {
```
This is actually dead code (became obsolete after improving the master fail-over handling in a recent PR).  `SnapshotsService#processExternalChanges` will remove any such snapshot entry without taking any additional action these days (since 8.x never creates such a snapshot, it can't have listeners for it so quietly dropping it from the cluster state is just fine).

original-brownbear committed:
drop dead code

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -257,4 +361,26 @@ public long repositoryStateId() {
+
+    public enum State {
+        WAITING((byte) 0),
+        META_DATA((byte) 1);
```
done :)

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -425,7 +425,6 @@ public void updateState(ClusterState state) {
 
     private long bestGeneration(Collection<? extends RepositoryOperation> operations) {
         final String repoName = metadata.name();
-        assert operations.size() <= 1 : "Assumed one or no operations but received " + operations;
```
Good question, this is a stupid situation that is the result of the logic we introduced around `org.elasticsearch.repositories.blobstore.BlobStoreRepository#bestEffortConsistency` that allows us more leniency when reading `RepositoryData` until we have finalized the first snapshot in the repository.
I don't think we should allow multiple operations with different generations for a single repository in the CS. It just doesn't make any sense and could even cause repository corruption when mounting an S3 repo with existing content (the operation with the lower gen id is clearly broken in this scenario).
=> maybe it's helpful to talk about this in the meeting later, but looking at this now I don't see a valid reason to have more than one value here per repository and I'd just implement logic that would prevent us from ever getting into this state.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                                    snapshotEntries.add(entry);
+                                }
+                            }
+                            updatedSnapshotsInProgress = new SnapshotsInProgress(snapshotEntries);
```
We are asserting correctness of the state on the CS applier thread in `SnapshotsService` anyway, that's why I generally stayed away from adding individual assertions on all CS updates but I can move the assertions into the tasks as well for easier debugging if you want?

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1433,4 +1943,16 @@ protected ClusterBlockException checkBlock(UpdateIndexShardSnapshotStatusRequest
+
+        private final SnapshotsInProgress.Entry entry;
+
+        private final Metadata metadata;
```
> We would have to write the metadata prior to enqueuing here, right?

Yea that would be one option. I like it in theory, but in practice I'd rather not do that as it makes the whole logic much more complex.

First off, I wonder if this all matters much in practice:

If the snapshot finalizations come in, in a spike then the metadata should be almost the same across all snapshots and with the way we deserialize should be mostly the same objects shouldn't it? I'm having a hard time imagining a situation where finalizations would massively queue up **and** have different metadata. You could have a situation where a snapshot abort or a data node dropping out causes a bunch of concurrent snapshots to fail and get queued up for finalization, but in that case they all come from the same CS update and have the same metadata object. Outside of that, a finalization should generally be as fast or faster than a shard snapshot (I guess there's some strange corner cases with mixed data and master nodes where running shard snapshots make finalization slow for example but even in that case I wouldn't expect those metadata objects to take up a lot of dedicated heap?).


Second of all, I mainly added the metadata to the queue to make things "subjectively more consistent" by reducing the skew between data and metadata snapshot. Technically we don't give any guarantees as to the time of when we take the metadata snapshot (only that it's after all shards have finished).  Maybe it would be enough to simply hold on to the latest metadata seen when movign a snapshot to `SUCESS` state and use that for all the snapshots we finalize. No strong need to too keep a copy per snapshot?

ywelsch replied:
> Maybe it would be enough to simply hold on to the latest metadata seen when movign a snapshot to SUCESS state and use that for all the snapshots we finalize. No strong need to too keep a copy per snapshot?

Wouldn't that also be the behavior when we  have a master failover (where the new master now uses a new cluster state as finalization point?

original-brownbear replied:
Yea exactly ... good point. I'll get rid of this. This "felt" smarter than it is ...

original-brownbear committed:
stop double wrapping list

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -466,10 +487,25 @@ public static State fromValue(byte value) {
+        for (Entry entry : entries) {
+            for (ObjectObjectCursor<ShardId, ShardSnapshotStatus> shard : entry.shards()) {
+                final ShardState shardState = shard.value.state();
+                if (shardState == ShardState.INIT || shardState == ShardState.ABORTED) {
```
I added `org.elasticsearch.cluster.SnapshotsInProgress.ShardSnapshotStatus#isAssigned` now as a method to also cover the waiting-for-shard-relocation case here that we also need to track here. I'll update the name of the map accordingly

original-brownbear committed:
rename

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -202,37 +228,54 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
+                final List<IndexId> indexIds = repositoryData.resolveNewIndices(
+                        indices, runningSnapshots.stream().filter(entry -> entry.repository().equals(repositoryName))
+                                .flatMap(entry -> entry.indices().stream()).distinct()
+                                .collect(Collectors.toMap(IndexId::getName, Function.identity())));
```
> Do we have a test somewhere that ensures that we're not resolving to new indices for concurrent snaps? (which would always work, but give us less incrementality)

I don't think this would work. Remember, we're looking up index uuids (the repo uuids) by index name. If we were to cause collisions there with multiple `IndexId` for a single name, then we would a. trip assertions in `RepositoryData` when adding a snapshot to it (updated shard generations would conflict with existing `RepositoryData`) and b. fail repository consistency checks. We do have a bunch of tests that run two or more concurrent snapshots on an empty repo that cover this properly I think (it was one of the first things that blew up when I enabled concurrent snapshots :)).

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
further cleanups

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1046,25 +1247,76 @@ public ClusterState execute(ClusterState currentState) {
+                final SnapshotsInProgress updatedSnapshots;
+
+                if (snapshots == null) {
+                    updatedSnapshots = new SnapshotsInProgress();
```
Gave this a go in https://github.com/elastic/elasticsearch/pull/58382

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
much cuter

original-brownbear committed:
fix

original-brownbear committed:
moar docs, cleaner logic

original-brownbear committed:
tweaks

original-brownbear committed:
fix

original-brownbear committed:
test cleanups

original-brownbear committed:
shorter

original-brownbear committed:
much more readable tests

original-brownbear committed:
much much stricter tests and fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
cleaner

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
even moar docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
test for and fix more listener leaks

original-brownbear committed:
better doc + naming

original-brownbear committed:
optimize

original-brownbear commented:
@ywelsch @tlrx I now way enhanced the documentation (and refactored logic to be more modular) and redid the tests a little to make this easier to review (make it more obvious what logical scenario each test covers).

original-brownbear committed:
writable

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -257,4 +361,26 @@ public long repositoryStateId() {
         }
     }
+
+    public enum State {
```
Sure done :)

original-brownbear committed:
faster

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -410,15 +410,22 @@ public IndexId resolveIndexId(final String indexName) {
             } else {
-                indexId = new IndexId(index, UUIDs.randomBase64UUID());
+                if (inFlightIds.containsKey(index)) {
+                    indexId = inFlightIds.get(index);
```
Sure done

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -113,6 +117,8 @@
+
     public static final Version SHARD_GEN_IN_REPO_DATA_VERSION = Version.V_7_6_0;
 
     public static final Version INDEX_GEN_IN_REPO_DATA_VERSION = Version.V_8_0_0;
```
Not yet backported sorry, incoming though (we discussed this last week, very tricky backport ...).

original-brownbear commented:
Sorry removed review requests for a bit, turns out GH hid some comments from me that I need to address first. Will re-request shortly

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1616,6 +1615,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                if (entry.repositoryStateId() == oldGen) {
+                    deletionEntries.add(entry.withRepoGen(newGen));
+                } else {
+                    deletionEntries.add(entry);
```
Currently, this would all be situations where we concurrently run stuff before the repo has written anything since it was mounted (so some ops pick up a broken generation, others pick up a good one). All of these situations are somewhat safe from corruption thanks to the fact that the finalization step of writing a new index-N will break for them (at least as safe as the current single operation situation as far as I can tell). We can't really assert any hard truths here I'm afraid (S3 could technically always find some index-N with a weird N at any point ... theoretically).

I mean technically we could also just not allow concurrent operations until after the first write to the repository to be in a spot that is easier to reason about. I rejected that idea to some degree, because:
1. The fact that concurrent operations would actually do extra listings for the last index-N in the unsafe startup scenario makes it more likely that we catch a broken operation. 
2. It requires more logic that mixes repo and snapshot service concerns in a confusing way.

Hope that helps, same as with the other comment on this. I'm happy to spell out all the scenarios I can think of for this here if desired (it's quite a few) :)

original-brownbear committed:
different exceptions

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -202,37 +228,54 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
+                if (repositoryData.getSnapshotIds().stream().anyMatch(s -> s.getName().equals(snapshotName)) ||
+                        runningSnapshots.stream().anyMatch(s -> {
+                            final Snapshot running = s.snapshot();
+                            return running.getRepository().equals(repositoryName)
```
Sure done

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1616,6 +1615,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     * @param newGen new safe repository generation
+     * @return updated cluster state
+     */
+    private ClusterState updateRepositoryGenerations(ClusterState state, long oldGen, long newGen) {
```
Sure probably better :)

original-brownbear committed:
renaming

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
simpler listener

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -677,19 +795,44 @@ private static boolean removedNodesCleanupNeeded(SnapshotsInProgress snapshotsIn
+    private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata, @Nullable RepositoryData repositoryData) {
+        final boolean newFinalization = endingSnapshots.add(entry.snapshot());
+        final String repoName = entry.repository();
+        synchronized (currentlyFinalizing) {
```
This synchronization on `currentlyFinalizing` in a bunch of spots is not great admittedly but for step 1 I traded efficiency for safety in many spots like this one.
I have some planned simplifications that will move all the logic for triggering deletes and finalizations exclusively to the master update thread (most of them already are) which will remove the need for most synchronization on the fields in `SnapshotsService`.

original-brownbear commented:
@ywelsch @tlrx alright, now I think I got to all the review comments (including the tricky hidden ones :)). This is ready for another pass, :crossed_fingers: the added documentation and better tests make this easier to follow now.

tlrx reviewed (commented):
I had a first look and saw nothing compromising. However, I don't think I've digested everything yet, so I'll give it another try tomorrow.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
resolve some conflicts

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

ywelsch commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/ConcurrentSnapshotsIT.java:
```diff
@@ -0,0 +1,971 @@
+import static org.hamcrest.Matchers.not;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class ConcurrentSnapshotsIT extends AbstractSnapshotIntegTestCase {
```
Should we also test for the situation where a repository is mounted multiple times under different name, but same path, and check that the repo does not become corrupted when concurrently accessed?

original-brownbear replied:
We could but the result won't hold up for S3 anyway in edge cases (and we have checks for concurrent repo modification already e.g. [here](https://github.com/elastic/elasticsearch/blob/master/server/src/internalClusterTest/java/org/elasticsearch/snapshots/CorruptedBlobStoreRepositoryIT.java#L93)). I wonder if this is a good time to forbid mounting multiple writable repositories to the same path instead? (it's not too hard to do this on a best-effort basis I think and should prevent corruption issues from this kind of scenario in pretty much all realistic scenarios)

ywelsch replied:
I wonder if instead of forbidding, different repo definitions pointing to the same path should be treated the same way as if they were the same repo (i.e. concurrency limitations). Perhaps a first overapproximation would be to treat every repo as if it was the same (i.e. remove the "isSameRepo" checks in this PR).

Most users are making use of one repository, not many. They get concurrency benefits with this PR even if it treats every repo as being the same. We can open the system to more concurrency possibly later then once we have a stronger notion of "same repo"?
WDYT?

ywelsch reviewed (commented):
Weekend, I couldn't get further than this today.

original-brownbear committed:
changes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
fix corner case

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1614,6 +1613,45 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     * @param newGen new safe repository generation
+     * @return updated cluster state
+     */
+    private ClusterState updateRepositoryGenerationsIfNecessary(ClusterState state, long oldGen, long newGen) {
```
We're using `this.metadata`, found it nicer to use the field here than to pass in the name?

original-brownbear committed:
rename

original-brownbear committed:
add assertion

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -175,6 +175,16 @@ public Entry(Entry entry, ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards)
         }
 
+        public Entry withRepoGen(long newRepoGen) {
+            return new Entry(snapshot, includeGlobalState, partial, state, indices, dataStreams, startTime, newRepoGen, shards, failure,
```
Sure can, but it's not always `+1` (on master fail-over during finalization or delete we will skip a value of `N` here to avoid overwriting `index-N`)

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -404,6 +425,16 @@ public String reason() {
+         * ({@link ShardState#INIT} or {@link ShardState#ABORTED}) or is in state {@link ShardState#WAITING} with a concrete non-null
+         * node id assignment (i.e. waiting for a shard relocation/initialization to finish).
+         */
+        public boolean isAssigned() {
```
> Isn't it sufficient to look at shard states here that have an non-null node id?

As a result of me (re-)using the `WAITING` state with node id `null` and with node id `!= null` we can't do this. Might be worth while to go for a new kind of `ShardState` here after all. I'll do that before re-requesting review. On it :)

original-brownbear committed:
add assertion

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
add queued state for shard snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
50% fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
fixes

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
moar coverage

original-brownbear committed:
moar coverage

original-brownbear committed:
docs

original-brownbear committed:
even moar coverage

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
fix

original-brownbear committed:
cover another tricky corner case

original-brownbear committed:
and another corner case

original-brownbear committed:
almost full coverage

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
another failure handler tested

original-brownbear committed:
fix bugs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
docs and simpler

original-brownbear committed:
cover and fix very edgy edge case

original-brownbear commented:
@tlrx @ywelsch Alright :) finally all points addressed I think. Introduced a new shard snapshot state `QUEUED` to simplify things and got to pretty much 100% coverage of the logic now (just a few random error handlers left that seem impossible to hit in practice but are equivalent to "not master" hooks in code pretty much so not too worried there).

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
fix conflict

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1463,4 +2157,171 @@ protected ClusterBlockException checkBlock(UpdateIndexShardSnapshotStatusRequest
+
+        @Override
+        public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+            logger.trace("Removed all snapshot tasks for repository [{}] from cluster state, now failing listeners", repository);
```
should we log this at info level? This is an important event?

original-brownbear replied:
Yea let's even do `WARN` here. The only way to run into this is when failing to load `RepositoryData`. This should be almost impossible with retries and caching of `RepositoryData` and if it happens at any sort of frequency it should be very visible.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1463,4 +2157,171 @@ protected ClusterBlockException checkBlock(UpdateIndexShardSnapshotStatusRequest
+
+        @Override
+        public void onFailure(String source, Exception e) {
+            logger.debug(
```
should we log at info level? Important event, no?

original-brownbear replied:
Yea we could log this more visibly. It's just something that will happen on master-failover and I didn't want to addd to the general noise of that but that's not an actual argument => info it is :)

ywelsch reviewed (commented):
I've left some more comments. Getting close, though

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -134,6 +141,14 @@
+    private final Map<String, List<ActionListener<Void>>> snapshotDeletionListeners = new HashMap<>();
+
+    //Set of repositories currently running either a snapshot finalization or a snapshot delete.
+    private final Set<String> currentlyFinalizing = Collections.synchronizedSet(new HashSet<>());
```
We do check that by running the assertion `org.elasticsearch.snapshots.SnapshotsService#assertAllListenersResolved` after each IT or are you looking for a different kind of check?

ywelsch replied:
yeah, that only covers the listeners. I would like to cover also the  remaining  state here (i.e. snapshotDeletionListeners is empty, currentlyFinalizing is empty, ...)

original-brownbear replied:
We do check all of those in that method now https://github.com/elastic/elasticsearch/pull/56911/files#diff-a0853be4492c052f24917b5c1464003dR2012 ?

ywelsch replied:
ok, missed  that. All good.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1095,11 +1395,64 @@ public void onFailure(String source, Exception e) {
+                    if (tryEnterRepoLoop(repoName)) {
+                        deleteSnapshotsFromRepository(newDelete, repositoryData, newState.nodes().getMinNodeVersion());
+                    } else {
+                        logger.trace("Delete [{}] could not execute directly and was queued", newDelete);
```
Idk, this is super trivial and I mainly just added it for my debugging of tests. It's just that this is logged in the  corner case of starting a delete in the short window of time between the cluster state update that removed the latest snapshot for a given repo and the listener for that snapshot getting resolved. We can't do synchronisation there because we delete outdated index-N between the two events so this is how I solved it for now, but it's extremely unlikely to get here and completely irrelevant outside of tricky debugging.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1047,45 +1291,101 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                        return currentState;
+                    }
+                }
+                if (replacedEntry == null) {
```
Sorry this was a left-over from some previous iteration, obviously redundant condition -> removing.

original-brownbear committed:
CR: small fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
CR: add setting to limit concurrency

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -204,24 +221,34 @@ public ClusterState execute(ClusterState currentState) {
                 // cluster state anyway in #applyClusterState.
-                if (snapshots.entries().stream().anyMatch(entry -> entry.state() != State.INIT)) {
+                if (minNodeVersion.before(FULL_CONCURRENCY_VERSION)
+                        && runningSnapshots.stream().anyMatch(entry -> entry.state() != State.INIT)) {
```
Done in 4c0fcafb384

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1152,82 +1505,359 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+     * @param e exception that caused us to realize we are not master any longer
+     */
+    private void failAllListenersOnMasterFailOver(Exception e) {
+        synchronized (currentlyFinalizing) {
```
Done

original-brownbear commented:
> Perhaps a first overapproximation would be to treat every repo as if it was the same (i.e. remove the "isSameRepo" checks in this PR).

This would not be entirely trivial to do from where the code is at currently.

More importantly though, even if we did this, it still wouldn't in-fact be safe on S3 I think. With repository data caching taking  multiple snapshots to S3 (even if done on-by-one) from multiple write mounts seems to always be an issue (if the current index-N moves while we do snapshots we simply don't notice).

I thought about this a little and to me it seems the safest option we have might be to do the following things:

1. Merge https://github.com/elastic/elasticsearch/pull/57786 so that we never execute deletes if we would suspect concurrent modification from missing shard generations.
2. Do the same for snapshot create. If a data-node fails a snapshot due to being unable to read a given shard generation -> do not finalise the snapshot.

For both of these I would actually go a step further than the PR I linked and mark the repository as corrupted (like we do if we're not seeing an index-N that expect) in the CS if we run into a missing shard generation. 

3. I also (didn't do this yet but I think we have to for safety reasons on S3) think we should not allow any concurrent snapshots so long as there is any numeric shard generation in the repository.

I think we strictly enforce these 3 points we don't need any further constraints on concurrency and with the exception of extreme corner cases on S3 (there's simply limits to what we can do with eventual consistency here even in theory) we should be safe no matter what the user does.

... as a matter of fact, I think we should do the above 3 things in any case because they prevent corruption from concurrent access by multiple clusters as well.

... I'm happy to discuss this on another channel as well. This is a bit of tricky topic that relates to caching repository metadata heavily. The more we cache and don't do physical reads of the repo contents, the more possible scenarios of corrupting a repository due to concurrent access we invite (at least on S3, but also on other repo types in some cases) and it may be good to come to an understanding of the trade-offs we're willing to make here?

original-brownbear commented:
Thanks so much @ywelsch I know this is a tricky one! I think I addressed all points again, but I think there's some discussions to be had around the multiple repo mounts situation still. See (https://github.com/elastic/elasticsearch/pull/56911#issuecomment-653591443)

original-brownbear commented:
Discussed this with @ywelsch on another channel and we determined a few new dependencies to this PR:

1. We decided to fail operations on missing shard level metadata as suggested in https://github.com/elastic/elasticsearch/pull/56911#issuecomment-653591443
  * This entails adding a way of repairing a repository though since we don't feel comfortable with simply marking a repository as corrupted without giving the user a way out of the situation other than moving on to a new repository. I will start work on a solution for this shortly.
2. We will add a check to snapshot finalization that makes sure that the `index-N` blob that a finalization is based on actually exists to avoid a class of possible concurrent repo access bugs
3. We want to add a limit to the number of concurrently snapshotting shards in place of the limit on concurrent repository operations to limit memory use. Limiting the concurrent operations could be trouble for ILM that might start running snapshots for a large number of indices expecting mechanics similar to those of moving shards. (will do in this PR)
4. In case a repository does not yet use uuid-shard generations in all its snapshots we want to queue up snapshot operations (instead of the more elaborate but in this case unsafe concurrency) to not create inconsistent SLM behavior across versions.

That's all let me know if I missed anything Yannick :) I'm on it working through the list

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
merge in master

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
merge in master

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge branch 'master' of github.com:elastic/elasticsearch into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
add test for concurrent ops and parallel ops

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
be smarter with shard generations to always allow for concurrent stuff

original-brownbear commented:
I worked through this list again:

> add a check to snapshot finalization that makes sure that the index-N blob that a finalization is based on actually exists to avoid a class of possible concurrent repo access bugs

Done in https://github.com/elastic/elasticsearch/pull/59141
IMO this provides about as much safety around concurrent repository access (from multiple mounts or otherwise) as we will get. On consistent blob stores I think it fixes all situations. On S3 it makes it incredibly unlikely that concurrent mounts would cause trouble as well. Keep in mind, even if you do run concurrent writes for concurrent repo mounts to the same location, the repo will get marked as corrupted in one of the two mounts as soon as an index-N blob is missing unexpectedly. Since the logic around writing new index-N blobs involves as CS update before and after writing and index-N blob, these writes are spaced out a little even across multiple repositories and approximately serialized via the cluster state update thread, making it even less likely to run into trouble on S3.

> add a limit to the number of concurrently snapshotting shards in place of the limit on concurrent repository operations to limit memory use. Limiting the concurrent operations could be trouble for ILM that might start running snapshots for a large number of indices expecting mechanics similar to those of moving shards. (will do in this PR)

I though about this some more. I don't think this is necessary. I think we can simply deduplicate `ShardId` and `Index` instances when deserializing `SnapshotsInProgress#Entry` instances to lower memory use for `SnapshotsInProgress` but beyond that I'm not sure how much of a concern this really is. Maybe we should just put an upper bound of `1000` concurrent operations by default (as that is pretty much the upper bound for a reasonable number of snapshots per repo that we've been communicating in the past) to limit things in case of broken SLM configurations and that's it for this?

> In case a repository does not yet use uuid-shard generations in all its snapshots we want to queue up snapshot operations (instead of the more elaborate but in this case unsafe concurrency) to not create inconsistent SLM behavior across versions.

I think this is actually not necessary in hindsight :) I pushed https://github.com/elastic/elasticsearch/pull/56911/commits/f5a44d4bfa3d8bd5940cc09f694ec3bc99b15b1d to force the use of shard generations even if it's numeric ones for queued up shard level snapshots. Meaning, even if your shard level generations are numeric queued up shard level snapshots won't do listing to find the next generation and hence are safe even on S3. This change will need to be a adapted a little for the backport to 7.x because there are some corner cases here in mixed version clusters (literally only with `7.7.0` I think) where we want to filter out numeric generations in the blob store still but in mixed version clusters with `7.8.0+` it's safe as far as I can see. Certainly, it seems to me that this is the safer approach compared to adding a different state machine for queuing operations while the repository is at an older version.

=> IMO this is ready for another review.

ywelsch reviewed (approved):
LGTM. Thanks for tackling this challenge, and good luck with the backport.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear committed:
CR: limit 1k

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -438,16 +477,34 @@ public static State fromValue(byte value) {
+            try {
+                assert assertConsistentEntries(entries);
+            } catch (AssertionError e) {
+                throw e;
```
I'm not sure to understand why we catch and rethrow here

original-brownbear replied:
So I could put a debug breakpoint there :D  Thanks for spotting!

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -142,6 +158,18 @@
+    private final OngoingRepositoryOperations repositoryOperations = new OngoingRepositoryOperations();
+
+    /**
+     * Setting that specifies the maximum number of allow concurrent snapshot create and delete operations in the
```
allow -> allowed

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -204,26 +235,36 @@ public ClusterState execute(ClusterState currentState) {
+                        currentState.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY);
+                if (deletionsInProgress.hasDeletionsInProgress() && concurrentOperationsAllowed == false) {
                     throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName,
                         "cannot snapshot while a snapshot deletion is in-progress in [" + deletionsInProgress + "]");
```
When backporting, we could maybe indicate in the error message that concurrent snapshot/deletions are possible in version 7.9?

original-brownbear replied:
:+1: Right, I put down a note for that when doing the back-port work.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -204,26 +235,36 @@ public ClusterState execute(ClusterState currentState) {
+                if (concurrentOperationsAllowed == false && runningSnapshots.stream().anyMatch(entry -> entry.state() != State.INIT)) {
                     throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
                 }
+                ensureBelowConcurrencyLimit(repositoryName, snapshotName, snapshots, deletionsInProgress);
```
What happen if multiple snapshot operations are started but the maxConcurrentOperations settings is updated to a value lower than the current number of concurrent ops? Would it still be possible to enque more ops?

original-brownbear replied:
> What happen if multiple snapshot operations are started but the maxConcurrentOperations settings is updated to a value lower than the current number of concurrent ops? 

Existing ops won't be affected but you can't start new ones.

> Would it still be possible to enque more ops?
No then the ops have to come down to below the new limit first before we can enqueue more.

tlrx reviewed (approved):
LGTM This is great work, thanks Armin! I carefully reviewed most parts but I must admit that I lightly reviewed the "repository loop" part. 

I deeply apologize for the time it took me to review this PR. The reviewing experience was not great for me due to the amount of code changes, I think using a dedicated branch here would have made sense.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-multiple-snapshots

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/ConcurrentSnapshotsIT.java:
```diff
@@ -0,0 +1,1313 @@
+        ensureStableCluster(3);
+
+        awaitNoMoreRunningOperations();
+        expectThrows(RepositoryException.class, deleteFuture::actionGet);
```
Not really, it's just "failed to update repository". It's all in the cause here, but that's also just a JSON parse failure.

original-brownbear committed:
review comments

original-brownbear commented:
Thanks so much @ywelsch @tlrx !!!
